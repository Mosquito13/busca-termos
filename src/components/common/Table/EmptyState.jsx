import { useEffect, useState } from 'react';
import classNames from 'classnames';

import { ReactComponent as EmptyStateImage } from '../../../assets/svg/emptystate.svg';
import Heading4 from '../Heading4';

const EmptyState = () => {
  const [showImage, setShowImage] = useState(false);

  useEffect(() => {
    const imageTimeout = setTimeout(() => setShowImage(true), 3000);

    return () => {
      clearTimeout(imageTimeout);
      setShowImage(false);
    };
  }, []);

  const imageWrapperClasses = classNames(
    'table__empty-state-image-wrapper',
    showImage && 'table__empty-state-image-wrapper--open'
  );

  return (
    <div className="table__empty-state">
      <div className="table__empty-state-message">
        <Heading4 value="Nenhum resultado encontrado." />
      </div>
      <div className={imageWrapperClasses}>
        <div className="table__empty-state-image">
          <EmptyStateImage />
        </div>
      </div>
    </div>
  );
};

export default EmptyState;
