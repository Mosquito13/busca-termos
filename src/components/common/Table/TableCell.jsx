import PropTypes from 'prop-types';
import classNames from 'classnames';

import CopyTextButton from '../CopyTextButton';

const SMALL = 'small';
const AUTO = 'auto';

const TableCell = ({ size, value, header, onSelect, selected, id, itemId, hasCopyButton, odd }) => {
  const classes = classNames(
    'table__cell',
    odd && 'table__cell--odd',
    header && 'table__cell--header',
    selected && 'table__cell--selected',
    size === SMALL && 'table__cell--small'
  );

  return (
    <div className={classes} onClick={() => onSelect?.(itemId, id, value)}>
      <span className="table__cell-text">{value}</span>
      {hasCopyButton && !header && (
        <div className="table__cell-copy-button">
          <CopyTextButton
            value={value}
            noMarginRight={size === SMALL}
          />
        </div>
      )}
    </div>
  );
};

TableCell.propTypes = {
  /**
   * Valor a ser exibido
   */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * Tamanho da celula
   */
  size: PropTypes.oneOf([SMALL, AUTO]),
  /**
   * Indica se faz parte do cabecalho da tabela
   */
  header: PropTypes.bool,
  /**
   * Funcao executada ao selecionar uma celula
   */
  onSelect: PropTypes.func,
  /**
   * Indica se esta celula selecionada
   */
  selected: PropTypes.bool,
  /**
   * Identificador unico da celula
   */
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * Identificador unico do registro (linha)
   */
  itemId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * Indica se e linha impar (para usar a zebra)
   */
  odd: PropTypes.bool,
  /**
   * Indica se exibe botao de copia na celula
   */
  hasCopyButton: PropTypes.bool
};

export { SMALL, AUTO };
export default TableCell;
