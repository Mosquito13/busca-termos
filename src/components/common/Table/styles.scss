@import '../../../sass/styles.scss';

.table {
  display: flex;
  flex-direction: column;
  height: 100%;
  border-radius: $radius;
  overflow: hidden;
  user-select: none;
  @include themify() {
    border: 1px solid themed('color-border');
  }

  &__header {
    flex: 0;
    @include themify() {
      border-bottom: 1px solid themed('color-border');
    }
  }

  &__body {
    flex: 1;
  }

  &__row {
    display: flex;
    height: 3rem;

    &--header,
    &--header:hover {
      @include themify() {
        background-color: themed('color-base-darker');
      }
      color: currentColor;
    }
  }

  &__cell {
    flex: 1;
    display: flex;
    align-items: center;
    overflow: hidden;
    padding-left: 1rem;
    position: relative;
    @include themify() {
      box-shadow: inset 0 0 5px 3px themed('color-base');
    }

    &:hover {
      @include themify() {
        background-color: themed('color-table-hover');
        color: themed('color-base');
      }
    }

    &--odd {
      @include themify() {
        background-color: themed('color-base-lighter');
        box-shadow: inset 0 0 5px 3px themed('color-base-lighter');
      }
    }

    &--small {
      flex: 0 0 8rem;
    }

    &--header,
    &--header:hover {
      @include themify() {
        background-color: transparent;
        box-shadow: none;
        color: themed('color-text');
      }
    }

    &--selected {
      @include themify() {
        background-color: themed('color-contrast');
        color: themed('color-base');
      }
    }
  }

  &__cell-text {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    cursor: default;
  }

  &__empty-state {
    height: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    position: relative;
  }

  &__empty-state-image-wrapper {
    position: absolute;
    left: 50%;
    bottom: 0;
    max-height: 50%;
    transform: translateX(-50%);
    opacity: 0;
    transition: opacity 1s;

    &--open {
      opacity: 1;
    }
  }

  &__empty-state-image {
    height: 25rem;
    filter: opacity(0.3);

    svg {
      width: 100%;
      height: 100%;
    }
  }
}
