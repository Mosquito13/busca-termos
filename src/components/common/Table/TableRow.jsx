import PropTypes from 'prop-types';
import classNames from 'classnames';

import TableCell, { SMALL, AUTO } from './TableCell';

const buildCells = (header, columns, data, onSelect, selectedCell, idField, odd) =>
  columns.map(({ title, dataKey, size }) => {
    const keyPrefix = header ? 'head_' : 'col_';
    const value = header ? title : data[dataKey];
    const cellId = `{${data?.[idField]}_${dataKey}}`;

    return (
      <TableCell
        key={`${keyPrefix}${dataKey}`}
        id={cellId}
        itemId={data?.[idField]}
        size={size}
        value={value}
        header={header}
        onSelect={onSelect}
        selected={selectedCell === cellId}
        odd={odd}
      />
    );
  });

const TableRow = ({
  header,
  columns,
  data,
  odd,
  selectedCell,
  onSelect,
  idField
}) => {
  const classes = classNames(
    'table__row',
    header && 'table__row--header'
  );

  return (
    <div className={classes}>
      {buildCells(header, columns, data, onSelect, selectedCell, idField, odd)}
    </div>
  );
};

TableRow.propTypes = {
  /**
   * Indica se a linha e do cabecalho
   */
  header: PropTypes.bool,
  /**
   * Dados para a linha
   */
  data: PropTypes.object,
  /**
   * Indica se e impar para usar a zebra
   */
  odd: PropTypes.bool,
  /**
   * Configuracao das colunas da tabela/linha
   */
  columns: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string,
      dataKey: PropTypes.string,
      size: PropTypes.oneOf([SMALL, AUTO])
    })
  )
};

export default TableRow;
