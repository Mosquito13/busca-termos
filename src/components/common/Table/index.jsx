import { useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import AutoSizer from 'react-virtualized-auto-sizer';
import { FixedSizeList } from 'react-window';
import { useDispatch } from 'react-redux';

import { SMALL, AUTO } from './TableCell';
import EmptyState from './EmptyState';
import TableRow from './TableRow';

import coreActions from '../../../actions/core';

import './styles.scss';

const isOdd = (index) => index % 2 !== 0;

const Table = ({ data, columns, idField }) => {
  const dispatch = useDispatch();
  const [selectedCell, setSelectedCell] = useState(null);

  const onSelect = useCallback((itemId, cellId, cellValue) => {
    window.navigator.clipboard.writeText(cellValue);
    setSelectedCell(cellId);
    dispatch(coreActions.setSelectedItemId(itemId));
  }, [dispatch]);

  return (
    <div className="table">
      <div className="table__header">
        <TableRow header columns={columns} />
      </div>
      <div className="table__body">
        {Boolean(!data?.length) && <EmptyState />}
        {Boolean(data?.length) && (
          <AutoSizer>
            {({ width, height }) => (
              <FixedSizeList
                width={width}
                height={height}
                itemSize={30}
                itemCount={data.length}
              >
                {({ index, style }) => (
                  <div style={style}>
                    <TableRow
                      data={data[index]}
                      columns={columns}
                      odd={isOdd(index)}
                      onSelect={onSelect}
                      selectedCell={selectedCell}
                      idField={idField}
                    />
                  </div>
                )}
              </FixedSizeList>
            )}
          </AutoSizer>
        )}
      </div>
    </div>
  );
};

Table.propTypes = {
  /**
   * Campo unico dos dados que para identificacao da linha
   */
  idField: PropTypes.string.isRequired,
  /**
   * Dados para exibicao
   */
  data: PropTypes.array,
  /**
   * Configuracao das colunas da tabela
   */
  columns: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string,
      dataKey: PropTypes.string,
      size: PropTypes.oneOf([SMALL, AUTO])
    })
  )
};

export { SMALL, AUTO };
export default Table;
