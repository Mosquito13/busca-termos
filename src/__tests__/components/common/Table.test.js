import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import * as reactRedux from 'react-redux';

import Table, { AUTO, SMALL } from '../../../components/common/Table';

const dataMock = [
  {
    uid: 1,
    id: 'id1',
    name: 'name1'
  },
  {
    uid: 2,
    id: 'id2',
    name: 'name2'
  },
  {
    uid: 3,
    id: 'id3',
    name: 'name3'
  }
];

const columnsMock = [
  {
    dataKey: 'id',
    title: 'Id',
    size: SMALL
  },
  {
    dataKey: 'name',
    title: 'Name',
    size: AUTO
  }
];

jest.mock(
  'react-virtualized-auto-sizer',
  () =>
    ({ children }) =>
      children({ height: 500, width: 500 })
);

describe(`${Table.name}`, () => {
  describe('tests', () => {
    let wrapper, dispatchMock;

    beforeAll(() => {
      dispatchMock = jest.fn();

      jest
        .spyOn(reactRedux, 'useDispatch')
        .mockImplementation(() => dispatchMock);
      jest.spyOn(reactRedux, 'useSelector').mockImplementation(() => false);

      Object.assign(global.navigator, {
        clipboard: {
          writeText: jest.fn()
        }
      });

      wrapper = mount(
        <Table idField="uid" data={dataMock} columns={columnsMock} />
      );
    });

    it('should render', () => {
      expect(wrapper.isEmptyRender()).toBeFalsy();
      expect(wrapper.find('[data-testid="empty-state"]')).toHaveLength(0);
      expect(wrapper.find('[data-testid="table-row"]')).toHaveLength(3);
      expect(
        wrapper.find('[data-testid="fixed-size-list"]').props().itemSize
      ).toBe(30);
    });

    it('should fire select', () => {
      act(() => {
        wrapper.find('[data-testid="table-row"]').at(1).props().onSelect();
      });
      wrapper.update();
      expect(dispatchMock).toBeCalledTimes(1);
    });

    it('should render with compact layout', () => {
      wrapper.unmount();
      jest.spyOn(reactRedux, 'useSelector').mockImplementation(() => true);

      wrapper = mount(
        <Table idField="uid" data={dataMock} columns={columnsMock} />
      );

      expect(wrapper.isEmptyRender()).toBeFalsy();
      expect(wrapper.find('[data-testid="empty-state"]')).toHaveLength(0);
      expect(wrapper.find('[data-testid="table-row"]')).toHaveLength(3);
      expect(
        wrapper.find('[data-testid="fixed-size-list"]').props().itemSize
      ).toBe(24);
    });

    it('should render empty state', () => {
      wrapper.setProps({ data: [] });
      expect(wrapper.isEmptyRender()).toBeFalsy();
      expect(wrapper.find('[data-testid="empty-state"]')).toHaveLength(1);
      expect(wrapper.find('[data-testid="table-row"]')).toHaveLength(0);
    });

    afterAll(() => {
      wrapper.unmount();
    });
  });

  describe('snapshots', () => {
    it('should render with default layout', () => {
      jest.spyOn(reactRedux, 'useDispatch').mockImplementation(() => jest.fn());
      jest.spyOn(reactRedux, 'useSelector').mockImplementation(() => false);

      const tree = renderer
        .create(<Table idField="uid" data={dataMock} columns={columnsMock} />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render with compact layout', () => {
      jest.spyOn(reactRedux, 'useDispatch').mockImplementation(() => jest.fn());
      jest.spyOn(reactRedux, 'useSelector').mockImplementation(() => true);

      const tree = renderer
        .create(<Table idField="uid" data={dataMock} columns={columnsMock} />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
