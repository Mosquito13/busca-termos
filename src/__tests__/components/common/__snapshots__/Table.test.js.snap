// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table snapshots should render with compact layout 1`] = `
<div
  className="table"
>
  <div
    className="table__header"
  >
    <div
      className="table__row table__row--header"
    >
      <div
        className="table__cell table__cell--header table__cell--small"
        onClick={[Function]}
      >
        <span
          className="table__cell-text"
        >
          Id
        </span>
      </div>
      <div
        className="table__cell table__cell--header"
        onClick={[Function]}
      >
        <span
          className="table__cell-text"
        >
          Name
        </span>
      </div>
    </div>
  </div>
  <div
    className="table__body"
  >
    <div
      onScroll={[Function]}
      style={
        Object {
          "WebkitOverflowScrolling": "touch",
          "direction": "ltr",
          "height": 500,
          "overflow": "auto",
          "position": "relative",
          "width": 500,
          "willChange": "transform",
        }
      }
    >
      <div
        style={
          Object {
            "height": 72,
            "pointerEvents": undefined,
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "height": 24,
              "left": 0,
              "position": "absolute",
              "right": undefined,
              "top": 0,
              "width": "100%",
            }
          }
        >
          <div
            className="table__row"
          >
            <div
              className="table__cell table__cell--small"
              onClick={[Function]}
            >
              <span
                className="table__cell-text"
              >
                id1
              </span>
            </div>
            <div
              className="table__cell"
              onClick={[Function]}
            >
              <span
                className="table__cell-text"
              >
                name1
              </span>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "height": 24,
              "left": 0,
              "position": "absolute",
              "right": undefined,
              "top": 24,
              "width": "100%",
            }
          }
        >
          <div
            className="table__row"
          >
            <div
              className="table__cell table__cell--odd table__cell--small"
              onClick={[Function]}
            >
              <span
                className="table__cell-text"
              >
                id2
              </span>
            </div>
            <div
              className="table__cell table__cell--odd"
              onClick={[Function]}
            >
              <span
                className="table__cell-text"
              >
                name2
              </span>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "height": 24,
              "left": 0,
              "position": "absolute",
              "right": undefined,
              "top": 48,
              "width": "100%",
            }
          }
        >
          <div
            className="table__row"
          >
            <div
              className="table__cell table__cell--small"
              onClick={[Function]}
            >
              <span
                className="table__cell-text"
              >
                id3
              </span>
            </div>
            <div
              className="table__cell"
              onClick={[Function]}
            >
              <span
                className="table__cell-text"
              >
                name3
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table snapshots should render with default layout 1`] = `
<div
  className="table"
>
  <div
    className="table__header"
  >
    <div
      className="table__row table__row--header"
    >
      <div
        className="table__cell table__cell--header table__cell--small"
        onClick={[Function]}
      >
        <span
          className="table__cell-text"
        >
          Id
        </span>
      </div>
      <div
        className="table__cell table__cell--header"
        onClick={[Function]}
      >
        <span
          className="table__cell-text"
        >
          Name
        </span>
      </div>
    </div>
  </div>
  <div
    className="table__body"
  >
    <div
      onScroll={[Function]}
      style={
        Object {
          "WebkitOverflowScrolling": "touch",
          "direction": "ltr",
          "height": 500,
          "overflow": "auto",
          "position": "relative",
          "width": 500,
          "willChange": "transform",
        }
      }
    >
      <div
        style={
          Object {
            "height": 90,
            "pointerEvents": undefined,
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "height": 30,
              "left": 0,
              "position": "absolute",
              "right": undefined,
              "top": 0,
              "width": "100%",
            }
          }
        >
          <div
            className="table__row"
          >
            <div
              className="table__cell table__cell--small"
              onClick={[Function]}
            >
              <span
                className="table__cell-text"
              >
                id1
              </span>
            </div>
            <div
              className="table__cell"
              onClick={[Function]}
            >
              <span
                className="table__cell-text"
              >
                name1
              </span>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "height": 30,
              "left": 0,
              "position": "absolute",
              "right": undefined,
              "top": 30,
              "width": "100%",
            }
          }
        >
          <div
            className="table__row"
          >
            <div
              className="table__cell table__cell--odd table__cell--small"
              onClick={[Function]}
            >
              <span
                className="table__cell-text"
              >
                id2
              </span>
            </div>
            <div
              className="table__cell table__cell--odd"
              onClick={[Function]}
            >
              <span
                className="table__cell-text"
              >
                name2
              </span>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "height": 30,
              "left": 0,
              "position": "absolute",
              "right": undefined,
              "top": 60,
              "width": "100%",
            }
          }
        >
          <div
            className="table__row"
          >
            <div
              className="table__cell table__cell--small"
              onClick={[Function]}
            >
              <span
                className="table__cell-text"
              >
                id3
              </span>
            </div>
            <div
              className="table__cell"
              onClick={[Function]}
            >
              <span
                className="table__cell-text"
              >
                name3
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
